app/tasks/schedule.php

```diff
@@ -0,0 +1,205 @@
+        $loadStart = \microtime(true);
+        /**
+         * Creating smaller functions list containing 5-min timeframe.
+         */
+        $timeFrame = DateTime::addSeconds(new \DateTime(), ENQUEUE_TIME_FRAME);
+        foreach ($functions as $function) {
+            $cron = new CronExpression($function['schedule']);
+            $next = DateTime::format($cron->getNextRunDate());
+            if ($next < $timeFrame) {
+                $queue[$next][$function['resourceId']] = $function;
+            }
+        }
+        $loadEnd = \microtime(true);
+        Console::success("Queue was built in " . ($loadEnd - $loadStart) . " seconds");
+    };
+
+    $removeFromQueue = function ($scheduleId) use (&$queue) {
+        foreach ($queue as $slot => $schedule) {
+            foreach ($schedule as $function) {
+                if ($scheduleId === $function['resourceId']) {
+                    Console::error("Unsetting :{$function['resourceId']} from queue slot $slot");
+                    unset($queue[$slot][$function['resourceId']]);
+                }
+            }
+        }
+    };
+
+    $dbForConsole = getConsoleDB();
+    $limit = 10000;
+    $sum = $limit;
+    $functions = [];
+    $queue = [];
+    $total = 0;
+    $loadStart = \microtime(true);
+    $latestDocument = null;
+
+    while ($sum === $limit) {
+        $paginationQueries = [Query::limit($limit)];
+        if ($latestDocument !== null) {
+            $paginationQueries[] =  Query::cursorAfter($latestDocument);
+        }
+        $results = $dbForConsole->find('schedules', \array_merge($paginationQueries, [
+            Query::equal('region', [App::getEnv('_APP_REGION')]),
+            Query::equal('resourceType', ['function']),
+            Query::equal('active', [true]),
+        ]));
+
+        $sum = count($results);
+        $total = $total + $sum;
+        foreach ($results as $document) {
+            $functions[$document['resourceId']] = [
+                'resourceId' => $document->getAttribute('resourceId'),
+                'schedule' => $document->getAttribute('schedule'),
+                'resourceUpdatedAt' => $document->getAttribute('resourceUpdatedAt'),
+            ];
+        }
+
+        $latestDocument = !empty(array_key_last($results)) ? $results[array_key_last($results)] : null;
+    }
+
+    $loadEnd = \microtime(true);
+    Console::success("{$total} functions where loaded in " . ($loadEnd - $loadStart) . " seconds");
+    $createQueue();
+    $lastUpdate =  DateTime::addSeconds(new \DateTime(), -FUNCTION_UPDATE_TIMER);
+
+    /**
+     * The timer updates $functions from db on last resourceUpdatedAt attr in X-min.
+     */
+    Co\run(
+        function () use ($removeFromQueue, $createQueue, $dbForConsole, &$functions, &$queue, &$lastUpdate) {
+            Timer::tick(FUNCTION_UPDATE_TIMER * 1000, function () use ($removeFromQueue, $createQueue, $dbForConsole, &$functions, &$queue, &$lastUpdate) {
```

Comment:
The timers do not look right 🤔 I don't see why they are embedded into another function on Line 111 ?

Reply:
Co\run(.  expects a callback  and using 2 timers without coroutine context issues an error
Deprecated: Swoole\Event::rshutdown(): Event::wait() in shutdown function is deprecated in Unknown on line 0